networkit/cpp/io/NetworkitBinaryWriter.cpp

```diff
@@ -0,0 +1,162 @@
+		buffer[0] = 0;
+	} else {
+		n = (64 - tlx::clz(value)) / 7;
+		buffer[0] = (1 << n) | (value << (n + 1));
+		value >>= 8 - (n + 1);
+	}
+
+	for(int i = 0; i < n; i++) {
+		buffer[i+1] = value & 0xFF;
+	   	value >>= 8;
+	}
+	return n + 1;
+}
+
+uint64_t NetworkitBinaryWriter::encodeZigzag(int64_t value) {
+	return (value << 1) ^ (value >> 31);
+}
+
+void NetworkitBinaryWriter::write(const Graph &G, const std::string& path) {
+
+	std::ofstream outfile(path, std::ios::binary);
+	Aux::enforceOpened(outfile);
+
+	bool directed = G.isDirected();
+	bool weighted = G.isWeighted();
+	uint64_t weightFormat = 0;
+	Header header = {};
+
+	auto calculateBaseOffset = [&] () {
+		return sizeof(Header);
+	};
+
+	auto calculateAdjOffset = [&] () {
+		size_t off = nodes * sizeof(uint64_t); //prefSum
+		off += (chunks - 1) * sizeof(uint64_t); //first vertex
+		return sizeof(Header) + off;
+	};
+
+	auto calculateWeightsOffset = [&] () {
+		size_t off = G.numberOfEdges() * sizeof(uint8_t); //adj arrays
+		off += (chunks - 1) * sizeof(uint64_t);		  // adj offsets
+		return sizeof(Header) + off + calculateAdjOffset();
+	};
+
+	auto setFeatures = [&] () {
+		header.features |= (G.isDirected() & DIR_MASK);
+		header.features |= ((weightFormat << WGHT_SHIFT) & WGHT_MASK);
+	};
+
+	auto setHeader = [&] () {
+		strncpy(header.magic,"nkbg000",8);
+		header.checksum = 0;
+		setFeatures();
+		header.nodes = nodes;
+		header.chunks = chunks;
+		header.offsetBaseData = calculateBaseOffset();
+		header.offsetAdjLists = calculateAdjOffset();
+		header.offsetAdjTranspose = 0;
+		header.offsetWeights = 0;
+	};
+
+	auto writeHeader = [&] () {
+
+		outfile.write(header.magic, 8);
+		outfile.write(reinterpret_cast<char*>(&header.checksum), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.features), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.nodes), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.chunks), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetBaseData), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetAdjLists), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetAdjTranspose), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetWeights), sizeof(uint64_t));
+	};
+
+	count prefSum = 0;
+	nodes = G.numberOfNodes();
+	if (nodes < chunks) {
+		chunks = nodes;
+		INFO("reducing chunks to ", chunks, " chunks");
+	}
+
+	// Compute first vertex of each chunk.
+	std::vector<uint64_t> firstInChunk;
+	firstInChunk.push_back(0);
+	uint64_t firstNode = 0;
+	for(uint64_t c = 1; c < chunks; c++) {
+		firstNode += (nodes/chunks);
+		firstInChunk.push_back(firstNode);
+	}
+	firstInChunk.push_back(nodes);
+
+	// Compute encoded size of arrays and store in vector.
+	uint64_t prevSize = 0; 
+	std::vector<size_t> adjOffsets;	//  pref sum of size encoded adj arrays
+	for(uint64_t c = 0; c < chunks; c++) {
+		uint64_t adj_size = 0;
```

Comment:
Variable is not used. But `adjSize` would probably be a better name for `prevSize`.